src/osd/PrimaryLogPG.cc

```diff
@@ -12813,6 +12813,24 @@ int PrimaryLogPG::prep_object_replica_pushes(
   ceph_assert(is_primary());
   dout(10) << __func__ << ": on " << soid << dendl;
 
+  if (soid.snap && soid.snap < CEPH_NOSNAP) {
```

Comment:
this is the same logic as in recover_missing(). The only other callers to recover_missing() and prep_object_replica_pushes() ensure they recover the head before the snap.

A way to simplify this would be to extract this from recover_missing() into maybe_kick_recovery(), so the logic is kept in one place.

Reply:
It seems that *recover_primary()* is the caller to recover_missing() too, So we need change the logic in recover_primary() if we extract this logic from recover_missing()..

Reply:
you're right, I misread that. nevermind